### ISSUE TITLE
Bug: Fresh install leads to Nginx 404 error due to wrong path in default site template

### ISSUE THREAD

**Issue by tlartaud** [Varying-Vagrant-Vagrants/VVV#2729]
### Are you using the latest stable or develop branch version of VVV?

No

### Is it a new VVV, or an existing VVV that used to work?

Existing, worked but now broken

### Did you use a CustomFile?

No (default)

### Whats the problem?

I'm using a VVV instance for years, which has always worked, and I needed to destroy it to move from VMware to vBox.
I did not change anything, but after I provisionned the VM, I got a Nginx 404 error (with a brand new design I never saw btw ^^).

To fix that, I had to fix a wrong path in `public_html/vvv-nginx-default.conf`, which is generated from the [official VVV's template](https://github.com/Varying-Vagrant-Vagrants/custom-site-template/blob/master/provision/vvv-nginx-default.conf) and restart Nginx.

The generated template includes this line:
```
root         "{vvv_path_to_site}{vvv_public_dir}";
```

while, I guess, it should be:
```
root         "{vvv_path_to_site}/{vvv_public_dir}";
```

> Notice the missing `/` character.
> The resulting path on the VM looks like `wordpress-localpublic_html` which is wrong.

After I modified this path, Nginx stopped complaining.

However, what I can't get out of my mind is the following:
How is that possible that this template, including this typo, worked for years and somehow now stopped working?
The only issue I faced while I re-provisionned again is about the new nginx signing key that expired last month. I fixed that from one of the VVV's latest commit, but is there an issue with an older Nginx version? How Nginx was actually able to reach out the correct path and display the site while the root path was incorrect and missing a `/`? This mystery is actually making me losing my mind ^^

### How do we reproduce it?

Just provision a new VVV instance by using the default site template in `config.yml` I guess ?
```
repo: https://github.com/Varying-Vagrant-Vagrants/custom-site-template.git
```

### What is the output of `vagrant status`

```shell
.
```

### Which Operating System are you using?

Microsoft Windows

### Which provider are you using?

VirtualBox 7

**Comment by tomjn** [comment]
> The only issue I faced while I re-provisionned again is about the new nginx signing key that expired last month. I fixed that from one of the VVV's latest commit, but is there an issue with an older Nginx version? How Nginx was actually able to reach out the correct path and display the site while the root path was incorrect and missing a /? This mystery is actually making me losing my mind ^^

This is perplexing though it's possible you had an out of date site template repo, e.g. if you'd made changes and `git` considered it to have working directory changes. A lot of the provisioner assumes `vvv_path_to_site` has no trailing slash

This is the commit that removed the slash: https://github.com/Varying-Vagrant-Vagrants/custom-site-template/commit/cf7c0c30d9a68709c8d28071f8cc6cb48210c688

**Comment by tomjn** [comment]
Does the site in question use a custom public directory?

**Comment by tlartaud** [comment]
Thank you for your really fast answer, this is appreciated.

> it's possible you had an out of date site template repo
I'm not sure, this VVV instance has been initiated years ago, but my `config.yml` file is just a standard one, using the default template from your git repository:
```
  wordpress-local:
    skip_provisioning: false
    allow_customfile: true
    description: "The default WPCA build environment - local.wordpress.test"
    repo: https://github.com/Varying-Vagrant-Vagrants/custom-site-template.git
    hosts:
      - local.wordpress.test
    custom:
      locale: fr_FR
      delete_default_plugins: true
      install_plugins:
        - query-monitor
        - disable-emails
        - log-emails
      wpconfig_constants:
        WP_DEBUG: true
        WP_DEBUG_LOG: true
        WP_DISABLE_FATAL_ERROR_HANDLER: true # To disable in WP 5.2 the FER mode
        WP_ENV: development
        SAVEQUERIES: true
        WP_DEBUG_DISPLAY: true
        SCRIPT_DEBUG: true
```

> This is the commit that removed the slash
This is interesting. If you removed the slash, you actually assume, I guess, that `vvv_public_dir`  includes a slash as its first character, right? So it should be `/public_html`, and not just `public_html`, right?

Because if I look at the other paths from the template file, they include a slash:
```
error_log    "{vvv_path_to_site}/log/nginx-error.log";
```

The problem is that the resulting output, for me, in the conf file generated on the guest, while I today destroyed and re-provisioned, was including `wordpress-localpublic_html` instead of `wordpress-local/public_html`.

> Does the site in question use a custom public directory?
The thing is, I did not modified the default structure. My default Wordpress installation path is simply `vagrant/www/wordpress-local/public_html` (`vagrant/` folder from the host ofc, else `srv/` from the guest).

I am still confused I must admit.

**Comment by tomjn** [comment]
I'd say adding the slash back in is the correct course of action, I suspect there's a reason it's worked despite that though, perhaps this is related to a box change? It's likely your transition has brought an Ubuntu upgrade along with it

**Comment by tlartaud** [comment]
> I'd say adding the slash back in is the correct course of action  

I'd say too, but shouldn't this be fixed into the official VVV's template? Or should I expect to face this issue again when reprovisioning? I mean, isn't the default template being overritten on each provision?

> I suspect there's a reason it's worked despite that though, perhaps this is related to a box change?  

I always used a VVV version that used the same box: `config.vm.box = 'bento/ubuntu-20.04'`. I only switched from VMware to vBox, but there shouldn't be any incidence about that regarding a path generated in a config file.

> It's likely your transition has brought an Ubuntu upgrade along with it  

I'd say: why this slash has actually been removed? Isn't it a typo actually?
If I look at the commit, the slash hasn't been removed everywhere ? So I suspect that the slash has been removed from the site template **only**, but that the config was still working for some reason.
```
PUBLIC_DIR_PATH="${PUBLIC_DIR_PATH}/${PUBLIC_DIR}"
```
While we can edit the config and set our own public_html dir, while the default does not include a slash, like stated in your docs, I don't see how the actual code in the template from your official repository could lead into the correct path:
```
| `public_dir`             | string | `public_html`              | Change the default folder inside the website's folder with the WP installation
```

Could you please confirm that the generated path into the nginx conf file on your guest actually includes a slash or not? If the path is actually correct? Sorry to ask... But that could help me to figure out why the generated path is incorrect and make sure that the futures re-provisionings work as expected.

**Comment by tlartaud** [comment]
I needed to confirm my fears: I re-provisionned the VM, and I faced this issue again: 404 error from Nginx (because the file is being overritten and re-generated again):
![image](https://github.com/user-attachments/assets/a99b76a7-d163-46a4-9960-73d57567adf8)

From the code i can read from the `vvv-init.sh`, this makes sense to me that this is happening.
```
PUBLIC_DIR=$(get_config_value 'public_dir' "public_html")
```
```
NCONFIG=$(vvv_site_template_search_replace_in_file "${VVV_PATH_TO_SITE}/provision/vvv-nginx-default.conf" "{vvv_public_dir}" "${PUBLIC_DIR}")
```

The `vvv_site_template_search_replace_in_file `  function is actually replacing the token `{vvv_public_dir}` from your [official site template](${PUBLIC_DIR}), by `${PUBLIC_DIR}` which defaults to `public_html`.
While this template includes this line `root         "{vvv_path_to_site}{vvv_public_dir}";`, which is missing a slash, the result of the config file is the following:

`sudo nano /etc/nginx/custom-sites/vvv-wordpress-local-41680070.conf`

```
server {
    listen       80;
    listen       443 ssl;

    server_name  local.wordpress.test;
    root         "/srv/www/wordpress-localpublic_html";

    # Nginx logs
    error_log    "/srv/www/wordpress-local/log/nginx-error.log";
    access_log   "/srv/www/wordpress-local/log/nginx-access.log";

    # This is needed to set the PHP being used
    set          $upstream php;

    ssl_certificate "/srv/certificates/wordpress-local/dev.crt";
    ssl_certificate_key "/srv/certificates/wordpress-local/dev.key";

    # Nginx rules for WordPress, rewrite rules, permalinks, etc
    include      /etc/nginx/nginx-wp-common.conf;



    location ~* \.(css|eot|gif|ico|jpeg|jpg|js|png|svg|tiff|tiff|ttf|webp|woff|woff2)$ {
        expires 100d;
    }
}

```

As you can see, `root         "/srv/www/wordpress-localpublic_html";` is a wrong path and is missing a slash.

Hope this makes more sense :)

PS: from the code, I see I can create my own `vvv-nginx-custom.conf` file, and fix the missing slash myself.... But... I'm wondering why this is hapening now while the slash has been removed from a commit created in 2020, and why I'm suddently the only one facing the effect of such an old potential typo. That's why I would appreciate if you could confirm that the path generated on your guest is also correct. This would be trully appreciated.

**Comment by tomjn** [comment]
Other lines in the nginx config assume there is no slash so I've created a PR to address this at https://github.com/Varying-Vagrant-Vagrants/custom-site-template/pull/64, if you can confirm all looks good I'll merge it.

Overall I'd prefer it didn't contain a slash, and enforce this in the provisioner script if there's inconsistency found

**Comment by tlartaud** [comment]
> Other lines in the nginx config assume there is no slash

Thank you for the confirmation, this will help!

> if you can confirm all looks good I'll merge it.

However, I can actually confirm that your commit will fix the issue, but I'm not actually able to confirm it is the real fix since I don't understand why it stopped working at some point.
More than that, I've rebuilt that VVV instance with some of my mates last month, and we didn't encounter that issue (while on linux and vbox). So I assume there's a fresh change from one of the package included in this VM (most probably, Nginx, but not sure).
My intuition tells me that therre is a typo in this file since a long time, but that it was still working for some reason, but I can't imagine why. I suspect an Nginx update, but it's just a suspicion and I might be wrong. I'm not sure any older Nginx version was able to fix a wrong path missing a slash by itself (that would be a huge vulnerability reported as a CVE, and I can't find it). It could also be something else, like a script adding back the missing slash in one of those two tokens, but I can't find it.

Also, I am for now the only one who reported this issue, and it would be nice to not break the provisioning for hundreds of people using VVV, if the real root cause comes from something we don't actually understand or know.

I think your PR actually respond to this issue, but I don't think this typo fix should be merged until we know why it suddenly caused an issue, because, in some cases, I don't know, it could maybe fix the problem for some (new VVV instances), and create an new issue for others ones (simple re-provisionings without apt-get update).

---

The real question is: why an Nginx config file including a wrong root path missing a slash was able to work for years?
That would help to find the real fix for that.
I'd say you can merge that PR, but I fear that this merge will cause an issue for anyone who will re-provision an existing VM, instead of provisioning it from scratch.

For now, I'll just edit the config file manually on the guest each time I'll re-provision the VM and restart Nginx, hoping someone will also report the issue in the future, to confirm.
Feel free to wait before merging so we can be sure about this issue :) 

Thank you!

**Comment by tomjn** [comment]
There have been some changes to how replacements work in recent releases, a quick check of 3.13 vs 3.12 and 3.11 should confirm if that's the case, provisioning under docker would lead to outright brokenness with replacements not happening or being replaced with blank values

**Comment by tlartaud** [comment]
> Overall I'd prefer it didn't contain a slash, and enforce this in the provisioner script if there's inconsistency found

Oh, I finally realized what you meant there. You'd prefer to not include a slash so that if the public dir is set to empty, you do not end up with an unexpected trailing slash? Right?

---

***Theorically:***

The `vvv-init.sh` already checks if the public dir is set to empty or not, and set `${PUBLIC_DIR_PATH}` accordingly:
```
PUBLIC_DIR_PATH="${VVV_PATH_TO_SITE}"
if [ ! -z "${PUBLIC_DIR}" ]; then
  PUBLIC_DIR_PATH="${PUBLIC_DIR_PATH}/${PUBLIC_DIR}"
fi
```

So, you could replace the tokens from that template to one single token, which will be replaced with the correct path, without any trailing slash at the end, even if the public dir is set to empty.

---

***Solution:***

So, in that case, I'd say that the correct fix is to replace, in `provision/vvv-nginx-default.conf`, this:
```
    root         "{vvv_path_to_site}{vvv_public_dir}";
```
by this:
```
    root         "{vvv_path_to_public_dir}";
```

And, to replace, in `provision/vvv-init.sh`, this:
```
  if [ -f "${VVV_PATH_TO_SITE}/provision/vvv-nginx-custom.conf" ]; then
    echo " * A vvv-nginx-custom.conf file was found"
    NCONFIG=$(vvv_site_template_search_replace_in_file "${VVV_PATH_TO_SITE}/provision/vvv-nginx-custom.conf" "{vvv_public_dir}" "${PUBLIC_DIR}")
  else
    echo " * Using the default vvv-nginx-default.conf, to customize, create a vvv-nginx-custom.conf"
    NCONFIG=$(vvv_site_template_search_replace_in_file "${VVV_PATH_TO_SITE}/provision/vvv-nginx-default.conf" "{vvv_public_dir}" "${PUBLIC_DIR}")
  fi
```
by this:
```
  if [ -f "${VVV_PATH_TO_SITE}/provision/vvv-nginx-custom.conf" ]; then
    echo " * A vvv-nginx-custom.conf file was found"
    NCONFIG=$(vvv_site_template_search_replace_in_file "${VVV_PATH_TO_SITE}/provision/vvv-nginx-custom.conf" "{vvv_path_to_public_dir}" "${PUBLIC_DIR_PATH}")
  else
    echo " * Using the default vvv-nginx-default.conf, to customize, create a vvv-nginx-custom.conf"
    NCONFIG=$(vvv_site_template_search_replace_in_file "${VVV_PATH_TO_SITE}/provision/vvv-nginx-default.conf" "{vvv_path_to_public_dir}" "${PUBLIC_DIR_PATH}")
  fi
```

---

***Incoming issues:***

However:
- BREAKING CHANGE: applying the fix above to `vvv-nginx-custom.conf` will break the provisioning for people who created their own custom file (because they would need to apply that change and modify their custom template too, in order to match the changes from the `vvv-init.sh ` file, which will be updated while re-provisioning)
- POTENTIAL BREAKING CHANGE: I am actually checking the code from an old stable version locally. I do not know if there's any new code from VVV which could manipulate the tokens from the template. If, in latest versions, you introduced a code into the main VVV stable repo, which is manipulating the tokens like the `vvv-init.sh` does, in the custom-site-template repo, this would introduce a breaking change for some people because you would change the tokens from a template from an external repository without being able to make sure that the VVV's repo who's manipulate those tokens match those changes

Hope this makes sense!

**Comment by tomjn** [comment]
The custom public path is a feature of the custom site template provisioner not the main VVV provisioner itself, `vvv_public_dir` doesn't exist outside of that, so perhaps making sure it's always prefixed with a slash is the ideal approach?

**Comment by tomjn** [comment]
This in the Nginx config might help:
```
    # Debug search replacements
    #
    # vvv_hosts:        "{vvv_hosts}"
    # vvv_path_to_site: "{vvv_path_to_site}"
    # vvv_public_dir:   "{vvv_public_dir}"
    # upstream:         "{upstream}"
    # LIVE_URL:         "{{LIVE_URL}}"
    # vvv_tls_cert:     "{vvv_tls_cert}"
    # vvv_tls_key:      "{vvv_tls_key}"
```

**Comment by tomjn** [comment]
On docker it resulted in this:

```
    # Debug search replacements
    #
    # vvv_hosts:        "one.wordpress.test"
    # vvv_path_to_site: "/srv/www/wordpress-one/"
    # vvv_public_dir:   "public_html"
    # upstream:         "php"
    # LIVE_URL:         ""
    # vvv_tls_cert:     "ssl_certificate "/srv/certificates/wordpress-one/dev.crt";"
    # vvv_tls_key:      "ssl_certificate_key "/srv/certificates/wordpress-one/dev.key";"
```

**Comment by tomjn** [comment]
A few updates later I'm seeing:

```
    server_name  one.wordpress.test;
    root         "/srv/www/wordpress-one/public_html";


    # Debug search replacements
    #
    # vvv_hosts:        "one.wordpress.test"
    # vvv_path_to_site: "/srv/www/wordpress-one/"
    # vvv_public_dir:   "public_html"
    # upstream:         "php"
    # LIVE_URL:         ""
    # vvv_tls_cert:     "ssl_certificate "/srv/certificates/wordpress-one/dev.crt";"
    # vvv_tls_key:      "ssl_certificate_key "/srv/certificates/wordpress-one/dev.key";"
    # PUBLIC_DIR_PATH:  "/srv/www/wordpress-one/public_html"
    # root:             "/srv/www/wordpress-one/public_html"

    # Nginx logs
    error_log    "/srv/www/wordpress-one//log/nginx-error.log";
    access_log   "/srv/www/wordpress-one//log/nginx-access.log";

    # This is needed to set the PHP being used
    set          $upstream php;

    ssl_certificate "/srv/certificates/wordpress-one/dev.crt";
    ssl_certificate_key "/srv/certificates/wordpress-one/dev.key";
```

Lessons:
 - might it be useful to add these debug vars to the default nginx config
 - the trailing slash is present on docker, unsure why it wouldn't be there for you, we can enforce this from the VVV site provisioner for consistency though I worry this means it'll still be broken for people who didn't update
 - the slashes on the error logs etc have a double `//` which would need fixing
 - documentation updates?

**Comment by tomjn** [comment]
Also noting we can strip the trailing slash like this:

```
PUBLIC_DIR_PATH="${VVV_PATH_TO_SITE%/}"
if [ ! -z "${PUBLIC_DIR}" ]; then
  PUBLIC_DIR_PATH="${PUBLIC_DIR_PATH}/${PUBLIC_DIR}"
```

In my test I added a `{{PUBLIC_DIR_PATH}}`, note the double braces, to distinguish those from VVV and those from the `vvv-init.sh`, likewise variables that are replaced by VVV itself have `vvv_` prefix. I wonder if dumping these in a JSON file might help?

**Comment by tomjn** [comment]
also possibly useful https://gist.github.com/jmmitchell/a6eacafadefb5292c9abdadc8bab7852

<blockquote><img src="https://github.githubassets.com/assets/gist-og-image-54fd7dc0713e.png" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> Gist</div><div><strong><a href="https://gist.github.com/jmmitchell/a6eacafadefb5292c9abdadc8bab7852">One liner for adding or removing trailing slash of a string in Bash</a></strong></div><div>One liner for adding or removing trailing slash of a string in Bash - manipulate-trailing-slash.sh</div></blockquote>

**Comment by tlartaud** [comment]
> On docker it resulted in this:

```
# vvv_path_to_site: "/srv/www/wordpress-one/"
```

So, on Docker, we have a trailing slash at the end? I guess this is where the issue is. We might actually get a trailing slash at the end on some VVV versions, but not on some others, probably because the replacement tokens are generated differently, probably a breaking change has been introduced (eg: token format changed while the template did not).
```
    root         "{vvv_path_to_site}{vvv_public_dir}";

    # Nginx logs
    error_log    "{vvv_path_to_site}/log/nginx-error.log";
```

---

In any case, the [provision/vvv-nginx-default.conf](https://github.com/Varying-Vagrant-Vagrants/custom-site-template/blob/master/provision/vvv-nginx-default.conf) is actually wrong.

***Case 1: If `vvv_path_to_site` includes a traling slash at then end***
We'll end up with something like
```
    root         "/srv/www/wordpress-one/public_html";

    # Nginx logs
    error_log    "/srv/www/wordpress-one//log/nginx-error.log";
```
> We'll get a double slash for the error_log for instance. This is not critical, but actually not optimal. However, this will work and Nginx will not complain.

***Case 2: If `vvv_path_to_site` doesn't include a traling slash at then end***
We'll end up with something like
```
    root         "/srv/www/wordpress-onepublic_html";

    # Nginx logs
    error_log    "/srv/www/wordpress-one/log/nginx-error.log";
```
> We won't get a double slash for the error_log, but we'll get a missing slash for the root path. This is critical because Nginx will complain about a 404 error, because the path is wrong.

---

> Also noting we can strip the trailing slash like this:

This might work to make sure that `vvv_path_to_site` include a slash, depending on how the template is being fixed.
However, this fix should take care about people using VVV older's stable version. A fix like could possibly break their provisioning.

---

The Breaking change that messed up this:
I just cloned VVV to check the code, and checked out the develop brand, and searched in files for `vvv_path_to_site`. The issue instantly popped up in front of me.
This is what I can see in `provision/provision-site.sh`, while on the develop branch:
```
  VVV_PATH_TO_SITE=${VM_DIR} # used in site templates
  NCONFIG=$(vvv_search_replace "${NCONFIG}" "{vvv_path_to_site}" "${VM_DIR}/")
```
> Notice the trailoing slash at the end. The toekn is generated with a trailing slash at the end.

However, on my project instance, I am on stable 3.10.0, where it was, at that point, written like this:
```
  VVV_PATH_TO_SITE=${VM_DIR} # used in site templates
  local TMPFILE=$(mktemp /tmp/vvv-site-XXXXX)
  sed -i "s#{vvv_path_to_site}#${VM_DIR}#"  "${TMPFILE}"
```
> Notice that the token is generated without a trailing slash at then end.

---

So, I guess that the main issue comes from a breaking change from an update, that suddenly added a traling slash to the `vvv_path_to_site` token, which, in some cases, generates double slashes in the paths because the template hasn't been updated accordingly.

However, I guess that the main problem is that the template is actually expecting a `vvv_path_to_site` with a trailing slash at the end for `root` while it is waiting for a token without a trailing slash for others params such as `error_log`.

**Comment by tomjn** [comment]
hmm that change in `provision-site.sh` is a good find, this means adding `/` to the front of `public_dir` is the best approach. Eitherway we'd need to do some finessing in the site template to ensure the variables match our expectations regardless of what happened upstream

**Comment by tlartaud** [comment]
> this means adding / to the front of public_dir is the best approach

I do actually agree with that.

As you say, I'd say that the slash should be moved from `vvv_path_to_site` to `vvv_public_dir` in `provision-site.sh` and `vvv-init.sh`. Which means the actual PR you made on [provision/vvv-nginx-default.conf](https://github.com/Varying-Vagrant-Vagrants/custom-site-template/blob/master/provision/vvv-nginx-default.conf) isn't necessarry anymore, but that [provision/vvv-init.sh](https://github.com/Varying-Vagrant-Vagrants/custom-site-template/blob/master/provision/vvv-init.sh#L114) should be edited instead, to add this slash before `${PUBLIC_DIR}`:

```
  if [ -f "${VVV_PATH_TO_SITE}/provision/vvv-nginx-custom.conf" ]; then
    echo " * A vvv-nginx-custom.conf file was found"
    NCONFIG=$(vvv_site_template_search_replace_in_file "${VVV_PATH_TO_SITE}/provision/vvv-nginx-custom.conf" "{vvv_public_dir}" "/${PUBLIC_DIR}")
  else
    echo " * Using the default vvv-nginx-default.conf, to customize, create a vvv-nginx-custom.conf"
    NCONFIG=$(vvv_site_template_search_replace_in_file "${VVV_PATH_TO_SITE}/provision/vvv-nginx-default.conf" "{vvv_public_dir}" "/${PUBLIC_DIR}")
  fi
```

The good point is that this shouldn't create any breaking change, even for users who created a `vvv-nginx-custom.conf` file.

However, I am not good enough at bash scripting to send a PR, I can read it but not sure I can properly write it. Anyway, do not hesitate to ask me to confirm that the incoming PR for both those repositories are working as expected while on the develop or an old stable branch.

Side note: I said I provisionned the VVV instance last month without any issue (old stable branch), I verified, and I was wrong. We did that on the 11th June 2024, which was before [this commit](https://github.com/Varying-Vagrant-Vagrants/custom-site-template/commit/455286de1d55f22deb89cd80d04e548b41bab403#diff-aa8e7e2a700c7e66f92b1037f71c45488159db9cf81bd6145a48de26cc8f3c93L86), which, I guess, affected old VVV instances. This explains why it suddenly stopped working without any change being made by ourselves to our instance, which is actually loaded as a submodule into our project repository: When I reprovisionned last week, the site template `vvv-init.sh` got updated while the `provision-site.sh` wasn't.

**Comment by tomjn** [comment]
Clarifying this commit replace the search replace line:

https://github.com/Varying-Vagrant-Vagrants/VVV/commit/abea1f59de627e5276178d730ce1e61c74f6b10a

And in doing so added a trailing slash that wasn't previously there

**Comment by tomjn** [comment]
https://github.com/Varying-Vagrant-Vagrants/VVV/pull/2731 Removes the slash that was added and tries to remove any trailing slashes that might have made it in e.g. if the user added it in their config like this: `vm_dir: path/`

**Comment by tomjn** [comment]
Site template PR:

https://github.com/Varying-Vagrant-Vagrants/custom-site-template/pull/65

Tries to enforce these assumptions and also exposes `PUBLIC_PATH_DIR` for use

**Comment by tomjn** [comment]
@tlartaud can you take a look, perhaps even test these?

Assumptions:

 - public dir value starts with a `/`, should never be a double slash
 - double slashes should not appear in Nginx configs
 - `VM_DIR` or the path VVV things the site is in should not end in a slash

**Comment by tlartaud** [comment]
Hello,

I have tested the update of the custom-site-template, branch `fix/slashes`, on an old stable VVV version (so, this means I tried that without the changes you made on the main repository of course).

This works as expected, I don't have any 404 error anymore, and this is the generated nginx config file:
```

# DO NOT EDIT!!
#
# If you want to make changes, copy this file to nginx-config-custom.conf or it'll be lost on provision.
#
# This is a "template" for the default Nginx config, that actual config gets created on provision.
server {
    listen       80;
    listen       443 ssl;

    server_name  local.wordpress.test;
    root         "/srv/www/wordpress-local/public_html";

    # Nginx logs
    error_log    "/srv/www/wordpress-local/log/nginx-error.log";
    access_log   "/srv/www/wordpress-local/log/nginx-access.log";

    # This is needed to set the PHP being used
    set          $upstream php;

    ssl_certificate "/srv/certificates/wordpress-local/dev.crt";
    ssl_certificate_key "/srv/certificates/wordpress-local/dev.key";

    # Nginx rules for WordPress, rewrite rules, permalinks, etc
    include      /etc/nginx/nginx-wp-common.conf;



    location ~* \.(css|eot|gif|ico|jpeg|jpg|js|png|svg|tiff|tiff|ttf|webp|woff|woff2)$ {
        expires 100d;
    }

}

# VVV Search Replace Variables for Debugging
#
# Core VVV search replacements:
# vvv_hosts:        "local.wordpress.test"
# vvv_path_to_site: "/srv/www/wordpress-local"
# vvv_public_dir:   "/public_html"
# upstream:         "php"
# vvv_tls_cert:     "ssl_certificate "/srv/certificates/wordpress-local/dev.crt";"
# vvv_tls_key:      "ssl_certificate_key "/srv/certificates/wordpress-local/dev.key";"
#
# Site template specific replacements:
#
# LIVE_URL:         ""
# PUBLIC_PATH_DIR:  "{{PUBLIC_PATH_DIR}}"
```

---

Just a question, shouldn't we be able to set the `branch` in the `config.yml` file? I had to hardcode the branch name in the VagrantFile where arguments are being passed to the `provision-site.sh` script.

---

I cannot yet confirm that the whole fix also work while on the VVV's `fix/vm_dir_slashes`, I did not try, but I can confirm that the custom-site-template changes fixed the issue on this old stable release.

**Comment by tomjn** [comment]
> Just a question, shouldn't we be able to set the branch in the config.yml file? I had to hardcode the branch name in the VagrantFile where arguments are being passed to the provision-site.sh script.

Yes! Alongside `repo:` etc put `branch: branchname`

> `# PUBLIC_PATH_DIR:  "{{PUBLIC_PATH_DIR}}"`

That should have been replaced 🤔

**Comment by tomjn** [comment]
Fixed! Should have been `PUBLIC_DIR_PATH`

**Comment by tomjn** [comment]
@tlartaud I've merged the site template changes everything should work on `develop` now without any branch changes to the site templates

**Comment by tlartaud** [comment]
@tomjn Thank you for the hard work made to solve this issue, this is really appreciated.
